### DISCUSSION TITLE
Named function parameters

### DISCUSSION THREAD

**Discussion by JakobBruenker** [ghc-proposals/ghc-proposals#654] [Ideas]
I thought it might be useful to have a place to discuss various potential solutions to named function parameters (aka labeled arguments, aka keyword arguments), after it came up in https://github.com/ghc-proposals/ghc-proposals/pull/267#issuecomment-2131195729

Motivation
---

In Haskell, we determine which argument in an application corresponds to which parameter of the function is by their order. This makes for concise syntax, especially in combination with currying. But there are several benefits to also offering named arguments:

- It's much easier to see which argument is which from reading the code
- Accidentally mixing up arguments of the same type happens much less if you have to also provide their name
- Currying relies on the first arguments being exactly the ones you want to supply - a syntax for named arguments could make it possible to partially supply any combination of arguments
- Lastly, many other languages (e.g. Python, Scala, OCaml) have support for named arguments, providing evidence of their usefulness

Current State
---

While Haskell does not natively support named arguments, there are libraries you can use, e.g. [named](https://hackage.haskell.org/package/named).

One issue with this is that while you can use it within your own projects, it's unlikely that any libraries you want to use will use a named argument library for their API, and even if they do, there is no way to enforce uniformity in terms of which library is used.

Design Space
---

There are a few questions one needs to answer in order to come up with a viable design:

1. **Should there be a new syntax for function argument definition? Or should it rely on existing syntax - and with existing functions?**

   - While it's tempting to say "if you have `quiz foo bar = "answer"`, just use `foo` and `bar` as names", this approach has a couple of issues:
     - If you have two function equations, they can have different and conflicting names for the arguments
     - Until now, those names in function definitions have not been part of the public API, and could be changed without that being a breaking change
     - It gets awkward if you have more complex patterns, like `quiz (fooL, fooR) bar = "answer"`
   - Completely new syntax would sidestep these issues, but would make the syntax even more complex
     - It could be advantageous to have a way to force users to use named rather than positional arguments with a function you define, cf. https://github.com/ghc-proposals/ghc-proposals/discussions/513
   - One option that seems promising to me is to use whatever syntax will be used for [retained quantification](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst#quantifiers), be it `foreach` or something more similar to the syntax used for `-XLinearTypes` (or anything else). Like `forall`, this new quantifier will also be dependent, meaning you have to be able to use it in the rest of the type signature. This means you need to be able to give it a name, so you can refer back to it. And if you simply don't use it in the rest of the type signature, you have essentially what we're looking for. In practice, this might look like
      ```haskell
      quiz :: foreach (foo :: (Bool, String)) (bar :: Int) -> String
      quiz foo bar = "answer"
      ```
      - You would still be able to use these arguments positionally
      - However, you would not be able to use names for arguments in applications to functions that use the traditional syntax
      - You could also extend this to work with `forall`, so you can apply type arguments in a named manner. In fact, this might be essential for this to feel cohesive. [Doing this was previously rejected](https://github.com/ghc-proposals/ghc-proposals/pull/267#issuecomment-2133704799) on grounds leaking information about the bound variable, and so here, too, we have the concern that changing it now makes things part of the public API that previously were not. However, this would not be without precedent: When `-XTypeApplications` was introduced, the *order* of `forall`-bound variables suddenly became part of the public API.

2. **What should the syntax of named function application be?**

   - Imitating record update syntax, we could go for something like `quiz [bar = 42, foo = (False, "a")]`
   - Similarly, you could use parentheses instead of square brackets
   - If we want to retain more of the look and feel of Haskell of arguments being separated by spaces, we could go for something like `quiz (bar = 42) (foo = (False, "a"))`.
   - All of these currently result in parse errors, and so would be backwards compatible

3. **Should you be able to mix named and positional arguments?**

- I don't currently see any obvious roadblocks in allowing this, but there easily could be. Examples, using one of the syntaxes above:
  - `quiz (bar = 42) (False, "a")`
  - `quiz (False, "a") (bar = 42)`
  - This appears to require inferring named parameters sometimes, e.g. `quiz (bar = 42) :: foreach (foo :: (Bool, String) -> String).` Not sure how this interacts with invisible arguments.
- Nonetheless, a first version without this would certainly already provide value

----

Please let me know if you have any comments or other ideas.

**Comment by jvanbruegge** [comment]
I think the most natural way is too add row types, and polymorphic records (which I am still working on the design for) and then use a record as function argument. Depending on the design (namely of there is one canonical order the user is allowed to depend on) it would also be possible to use a type class to allow to write a function that takes the arguments one by one (not a big fan of this).

However, it would be one or the other. Which I think is fair in a functional language, given that in other languages you do not curry and morally pass a tuple. A record is like a named tuple.

**Reply by JakobBruenker** [reply]
This does seem like a good approach; thanks for your efforts working on this.

> However, it would be one or the other. Which I think is fair in a functional language, given that in other languages you do not curry and morally pass a tuple. A record is like a named tuple.

It may be worth noting that e.g. in Scala you can do `val partiallyApplied = quiz(_, bar = 42)`, combining named arguments and partial application, while morally passing a tuple. But while being able to combine partial application with named arguments or an alternative like row types would be nice, I'd agree that going with a solution that doesn't provide this would not be the end of the world.

**Comment by goldfirere** [comment]
From my perspective, the best way to attack this is to start from OCaml. That is, OCaml is a language that is exceedingly similar to Haskell, and it has labeled arguments. (That's the OCaml term, which I'll use here, though without prejudice as to what we might call it in Haskell.)

Here are the salient points about OCaml's design (numbered only for easy reference):

1. **Types:** Instead of writing `int -> int -> int option`, we write `numerator:int -> denominator:int -> int option`. Spaces are allowed around the `:` but no one ever puts them there.
2. **Definitions:** Instead of writing `fun n d -> ...`, we write `fun ~numerator ~denominator -> ...` or `fun ~numerator:_ ~denominator:0 -> None` (allowing an arbitrary pattern after the `:`). In effect, the first syntax is an abbreviation for `fun ~numerator:numerator ~denominator:denominator -> ...`. No spaces are allowed between the `~`, the label, and the `:`, if present. (`~` is not allowed as a user-defined operator in OCaml.)
3. **Applications:** Instead of writing `div n d`, we write `div ~numerator:n ~denominator:d`, or if your variables already have the right names, `div ~numerator ~denominator`.
4. **Currying:** The expression `div ~numerator` has type `denominator:int -> int option`, as you might expect.
5. **Reordering:** You can always say `div ~denominator ~numerator`, which has identical semantics to `div ~numerator ~denominator`.
6. **Reordering and currying, together:** The expression `div ~denominator` has type `numerator:int -> int option`, quite conveniently. Evaluating the partial application `div ~denominator` never has any effects (because it is actually a lambda).
7. **Reordering at definition sites** is not allowed. The function `fun ~denominator ~numerator -> ...` has a different type than `fun ~numerator ~denominator -> ...`; there is no way short of eta-expansion to convert between the definitions.
8. **Compositionality:** Because OCaml is a curried functional language, it supports types like `x:int -> 'a`, where `'a` is a type variable. This variable might get expanded to, say, `x:float -> char`, meaning a function might have a type `x:int -> x:float -> char`. Duplicate labels are thus allowed. At application sites, arguments with the same label are treated positionally, with the first ~x being the `int` and a second `~x` being the `float`. This situation is generally avoided, but the language supports it just fine.
9. **Mixing labeled and unlabeled arguments:** An unlabeled argument is treated just like a labeled argument with a special, empty label. Because repeated labels are allowed and are treated positionally, this just works. One consequence of this is that if you have `f : int -> x:bool -> char`, you can write `f ~x` to get a function of type `int -> char`, using reordering. Accordingly, when mixing labeled and unlabeled arguments, the labeled ones typically go last, as this is more expressive.
10. **Inference:** Inference around labeled arguments is a bit tricky. It's fine at definition sites: the order of the arguments at the definition site tells you exactly what type the function has. At applications, there's a check whether the type of the function is already known. If it is, use that type, allowing function arguments to be reordered (as above); otherwise, infer the type of the function using the argument order as written. (The details here are subtle, and OCaml's type-checking algorithm is so different from Haskell's that we'd probably have to re-think much of this aspect of the design. In the context of Haskell -- but not OCaml -- we could also declare that we never allow applications to labeled arguments unless the type of the function is known, much like how we do for type applications.)
11. **Optionality:** Labeled arguments were added to OCaml after its initial versions. There was thus a desire to allow programs to opt-out of labels. OCaml thus supports calling labeled functions without their labels, in case a library exports a labeled interface but clients don't wish to use the new feature.

Now, for some opinion, having programmed with these labeled arguments:

12. They're pretty fantastic.
13. It's essential to allow punning -- e.g. having `~x` stand in for `~x:x` in both definitions and applications. This punning strongly encourages a uniform custom around variable names, and this is a Good Thing. Bonus points if there are records around and use the same names. (OCaml's field puns work identically to Haskell's.)
14. Reordering is nice, but not essential. I think I'd rather write `f ~arg1:_ ~arg2` to make the skipped argument explicit rather than just use reordering to invisibly produce the right function.
15. Because of reordering (which maybe Haskell might not adopt), function application becomes n-ary. (Normally, functional languages are originally understood by treating `f a b c` as `(((f a) b) c)`, also called binary application.) This actually dovetails nicely with Quick Look.
16. We might want to consider separating labeled arguments by something other than `->`, making n-ary application apparent also in the type. (Basically, it would mean that a bunch of arguments are tupled together, but without actually allocating a tuple at runtime.) Then we wouldn't have to worry about the possibility of repeated labels -- not that repeated labels cause any trouble in practice in OCaml.
17. Labeled arguments are so nice that we ended up recently adding labeled tuples. (This feature is in Jane Street's branch only; we'll be working with the OCaml maintainers to upstream over the coming months.) Labeled tuples are just what you think they are, and they are widely liked. They are most often used for labeling multiple return values from a function.
18. There is a growing desire for labeled return values, but we don't have a concrete design yet.
19. Syntactically, labeled arguments are a little awkward: the `~` is required in terms (both definitions and applications) but forbidden in types; and the `~label` is lexed as one lexeme. We'd probably have to do similar stuff in Haskell. Interesting, we might be able to use `~` just like OCaml, given that it's a no-op to mark a variable pattern as lazy (outside of `-XStrict`). Actually that probably doesn't work out so well, but I'm sure we could cobble together something.
20. OCaml has no support for type arguments in applications (interestingly, it has supported type arguments in definitions for some time). So it naturally doesn't support labeled type arguments. But it should, as should Haskell.
21. Despite my comment earlier that reordering is inessential (which I do not retract), it does encourage a nice little pattern. In the `base` library (a commonly used, batteries-included library of common functions/datatypes; OCaml's equivalent of Haskell's `base` is called `stdlib`), we have `map : 'a list -> f:('a -> 'b) -> 'b list`, with the function labeled. This allows `map ~f:(fun x -> x + 1) nums` as well as `map nums ~f:(fun x -> (* very long function here *))`. It's nice because for very short functions, I like it early. But for very long functions, I prefer to have the list before the function. With labeled arguments and reordering, we can have our cake and eat it, too.
22. Labeled arguments reduce the need for `newtype`s, because they make it harder to confuse two nearby things of the same type. `newtype`s are still sometimes great, but the possibility of using labeled arguments makes the bar just a bit higher for when you'd want to define a `newtype`. (OCaml has `newtype`s; they're called `[@@unboxed]` types.)

Bottom line for me: Haskell should just do this.

23. Punning. Without question. I'd be against this feature without punning.
24. Skip inference in applications. That is, treat labeled application just like polymorphic application: it works only when the type of the function is already known. This simplifies it all nicely.
25. Less sure about this one, but maybe just to be provocative: reflect the n-ary nature of application in the type. That is, design the types so that a block of labeled arguments is specified together, not separated with the traditional right-associative `->`. At most one argument in the group can be unlabeled. Require parentheses in applications to specify overapplication and underscores to specify underapplication.
26. Sketch out a design extending to labeled tuples and return values, just to make sure that the labeled argument design is forward compatible.
27. Prefer lighter syntax that perhaps relies on whitespace tricks over heavier syntax that's more regular. If we had to bracket every labeled argument, I'm not sure they'd be as popular. Syntax matters, especially for convenience features.

I hope this is helpful grist for the mill! I would consider designing and pushing for the feature myself, but I know I won't have the time to really see it through. But hopefully this will all inspire someone else to take this and run with it. I'll be cheering loudly from the sidelines.

**Reply by AntC2** [reply]
Thank you @goldfirere. For comparison, does OCaml feature an _extensible_ records system (with field labels)? A lot of your description seems like a poor man's [**] row-type. "Extensible" is what it needs for partial application. So I mean more than "labelled tuples". The records system must be able to 'merge' records providing their labels are disjoint. (I'm not sure whether that's what @jvanbruegge has in mind.)

>  repeated labels are allowed and are treated positionally

Errk! That sounds like an anti-feature.

[**] Apologies for the sexist language. How do I say that politically correctly?

**Reply by Ericson2314** [reply]
Yeah, I thought OCaml did have extensible record types (the O?). So why does it also have this? Accident of history?

**Reply by AntC2** [reply]
According to [the O'Caml manual](https://v2.ocaml.org/manual/expr.html#sss:expr-records), the only things counting as 'records' are pre-defined fixed set of label-names.

There appears to be a sub-typing mechanism (presumably "the `O`") such that a record with some extra fields (labels) can be used in code that mentions fewer labels. So not _extensible_/mergable nor can you even dynamically write an expression denoting a (anonymous) tuple with a bunch of label-value pairs. (I believe purescript can do the anonymous tuple bit.)

**Reply by goldfirere** [reply]
OCaml has two features that resemble traditional row-types, but neither hits it on the nose.

* **Polymorphic variants:** OCaml allows you to use an arbitrary identifier as a constructor that can be pattern matched against:

```ocaml
let thing1 = `ThisIsNotPreDeclared 5 'x' "hello"
let thing2 = `NeverWrittenBefore
let f thing = match thing with
  | `FreshIdent x y -> x + y
  | `ThisIsNotPreDeclared z -> z
```

The type system allows you to specify a list of such constructors (including their arguments) as a type, and these types allow you to say "some subset of these constructors" or "some superset of these constructors" or "exactly these constructors". There are explicit conversions among the types. The explicit conversions lift through covariant datatypes (like `list` but not like `ref`); the conversions are no-ops at runtime. This conversion mechanism is very similar to Haskell's `coerce`, except that it's directional.

Polymorphic variants don't quite work for labeled arguments, though: They are not erased at runtime. And they are not particularly efficient at runtime either. Another drawback is that the flexible typing sometimes gums up inference; I don't think they'd be ergonomic enough for use as labeled arguments.

* **Objects:** OCaml has an object system (that's the O), which essentially contains extensible records / row-types as a subset of its feature set. It's not worth giving a detailed explanation here. The problem is that, like polymorphic variants, the feature is just too big to be used for labeled arguments or tuples. The representations aren't erased at runtime, and the type inference is challenging.

-------------------

Some very committed folks at Jane Street have variously used both of these features to pretend to have extensible records or labeled tuples, but basically the colleagues of those folks have resented the choice. There's nothing wrong with doing it (if you don't care about performance), exactly, but it's just painful in practice.

**Reply by JakobBruenker** [reply]
Would you, at the point this is fully feature-complete, essentially want to have an additional dimension along which (most) quantifiers vary, labeled vs unlabeled?

**Reply by goldfirere** [reply]
Not quite. If we view "unlabeled" as a special case of "labeled", then there would be a new dimension, *label*. Though when we add this to the existing mix of ways quantifiers can vary, it does give me pause.

**Reply by Ericson2314** [reply]
When stuff is dependent, positional makes a lot more sense too. (Variable usages or lack thereof do at least allow for a partial order telescope, but we're getting fancy.)

**Comment by simonpj** [comment]
> Types: Instead of writing int -> int -> int option, we write numerator:int -> denominator:int -> int option

So the labels are really a first part of types?  That is I can write
```
f :: (foo:Int -> bar:Int -> Int) -> Bool
```
but I can only apply that `f` to a function that has appropriate labels in its type.   so I could not apply `f (+)`, say because the type of `(+)` does not have those labels.

This must get in the way of higher order programming doesn't it?

**Reply by goldfirere** [reply]
> This must get in the way of higher order programming doesn't it?

Not so much in practice. But OCaml is more comfortable with writing explicit lambdas than Haskellers would like. Examples: OCaml has no operator sections, and no standard function composition operator. If you want to pass a composition of functions as a higher-order argument, write `fun x -> f (g x)`. I don't love this aspect of the language so much, but it means that people don't grumble so much when they have to write the eta-expansion.

Also: the fact that this stuff is in types strongly encourages folks to develop consistent naming schemes. So these arise, and it actually all works quite well.

**Reply by AntC2** [reply]
> So the labels are really a first part of types?

(Presumably that should be "first-class part of types"(?) -- as opposed to Haskell's `data` labels, which are more decorative than first-class.)

My experience with labels being "first-class part of types" is the label-value pair ends up much more like a `newtype` (as Richard says/"consistent naming scheme", but avoiding the bureaucracy of declaring them at top level), with the advantage of bunching several label-value pairs together -- sounds very like what Richard describes at "labeling multiple return values from a function".

**Reply by goldfirere** [reply]
Yes, the label-value pair does feel a little `newtype`ish, indeed. And, yes, field names on records in OCaml are required, unlike in Haskell. The non-optionality of these names (overlooking the legacy mode OCaml has allowing ignoring labeled-argument labels) is a feature, not a bug. Yes, it leads to more typing, but also to clearer code and fewer mistakes.

**Reply by AntC2** [reply]
> the label-value pair does feel a little `newtype`ish

For that reason, I think labels should be spelled starting upper-case. (Even if we think of them at term level, they're more like constructors.) If we then want at term level to 'apply' the label or pun on it, that'll mean some specific syntactical prefix: Richard shows using `~x`; Hugs/Trex uses `#x`.

**Comment by VitWW** [comment]
Named Params should be similar to `Implicit Params` syntax, except it is not a part of Constraint:
```haskell
quiz :: (?foo :: (Bool, String)) -> (?bar :: Int) -> String
quiz ?foo ?bar = ... foo ... bar ...

baz = quiz (?bar = 42) (?foo = (False, "a"))
```
